server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -97,37 +101,41 @@ protected NodeRequest newNodeRequest(Request request) {
     }
 
     @Override
-    protected NodeGatewayStartedShards newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException {
-        final NodeGatewayStartedShards response = new NodeGatewayStartedShards(in, node);
+    protected NodeGroupedGatewayStartedShards newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException {
+        final NodeGroupedGatewayStartedShards response = new NodeGroupedGatewayStartedShards(in, node);
         assert response.getNode() == node;
         return response;
     }
 
     @Override
-    protected NodesGatewayStartedShards newResponse(
-        Request request,
-        List<NodeGatewayStartedShards> responses,
-        List<FailedNodeException> failures
-    ) {
-        return new NodesGatewayStartedShards(clusterService.getClusterName(), responses, failures);
+    protected NodesGroupedGatewayStartedShards newResponse(Request request,
+                                                           List<NodeGroupedGatewayStartedShards> responses,
+                                                           List<FailedNodeException> failures) {
+        return new NodesGroupedGatewayStartedShards(clusterService.getClusterName(), responses, failures);
     }
 
     @Override
-    protected NodeGatewayStartedShards nodeOperation(NodeRequest request, Task task) {
+    protected NodeGroupedGatewayStartedShards nodeOperation(NodeRequest request, Task task) {
+        NodeGroupedGatewayStartedShards groupedStartedShards = new NodeGroupedGatewayStartedShards(clusterService.localNode());
+        for (Map.Entry<ShardId, String> entry : request.getShards().entrySet()) {
```

Comment:
When sending these requests per-shard we execute them in parallel across the `FETCH_SHARD_STARTED` threadpool. I think we should continue to parallelise them at the shard level like that.